### PR TITLE
Add WebAssembly page.

### PR DESCRIPTION
**Pull request by jplevyak**
Signed-off-by: John Plevyak <jplevyak@gmail.com>

https://deploy-preview-6738--preliminary-istio.netlify.com/docs/concepts/wasm/

[ ] Configuration Infrastructure
[ X] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

File: content/en/docs/concepts/wasm/index.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
+---
+title: WebAssembly (Wasm) Plugins
+description: Describes Istio's WebAssembly Plugin system.
+weight: 25
+keywords: [wasm,webassembly,emscripten,extension,plugin,filter]
+---
+
+WebAssembly is a sandboxing technology which can be used to extend the Istio proxy (Envoy).  Going forward Istio extensions will be written as Envoy proxy plugins using the WebAssembly (Wasm) sandbox API. Mixer will be eliminated from the Istio architecture and all Mixer adapters must be eventually rewritten as proxy plugins.
```
Comment (PiotrSikora) [inline content/en/docs/concepts/wasm/index.md]: `s/Envoy proxy plugins/Proxy-Wasm plugins/`

File: content/en/docs/concepts/wasm/index.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
+---
+title: WebAssembly (Wasm) Plugins
+description: Describes Istio's WebAssembly Plugin system.
+weight: 25
+keywords: [wasm,webassembly,emscripten,extension,plugin,filter]
+---
+
+WebAssembly is a sandboxing technology which can be used to extend the Istio proxy (Envoy).  Going forward Istio extensions will be written as Envoy proxy plugins using the WebAssembly (Wasm) sandbox API. Mixer will be eliminated from the Istio architecture and all Mixer adapters must be eventually rewritten as proxy plugins.  
+
+WebAseembly sandbox Goal:
```
Comment (PiotrSikora) [inline content/en/docs/concepts/wasm/index.md]: `s/Goal/goals/`

File: content/en/docs/concepts/wasm/index.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
+---
```
Comment (PiotrSikora) [inline content/en/docs/concepts/wasm/index.md]: Throughout this document, we're referring to extensions as "plugins", "extensions" and "filters". We really need to pick one and stick to it.

File: content/en/docs/concepts/wasm/index.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+- **Operator** - An extension can be canaried and deployed as log-only, fail-open or fail-close.
+- **Extension developer** - The extensions can be written in several programming languages.
+
+This [video talk](https://youtu.be/XdWmm_mtVXI) is an introduction about architecture of WebAssembly integration.
+
+## High-level architecture
+
+Istio extensions have several components:
+
+- **Filter SPI** for building filters
```
Comment (PiotrSikora) [inline content/en/docs/concepts/wasm/index.md]: SPI?

File: content/en/docs/concepts/wasm/index.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+- **Sandbox** V8 embedded in Envoy
```
Comment (PiotrSikora) [inline content/en/docs/concepts/wasm/index.md]: `s/V8/V8 Wasm Runtime/`

File: content/en/docs/concepts/wasm/index.md (hunk 4 above)
Comment (jplevyak) [reply content/en/docs/concepts/wasm/index.md]: That was something that mandar had in his doc.  not sure.... @mandarjog

Comment (mandarjog) [reply content/en/docs/concepts/wasm/index.md]: it’s Service Provider Interface.

You implement an SPI and use an API.
If the name bothers you we can just use API everywhere.

File: content/en/docs/concepts/wasm/index.md (hunk 5 above)
Comment (jplevyak) [reply content/en/docs/concepts/wasm/index.md]: Done.

File: content/en/docs/concepts/wasm/index.md (hunk 4 above)
Comment (jplevyak) [reply content/en/docs/concepts/wasm/index.md]: Added alliteration.

File: content/en/docs/concepts/wasm/index.md (hunk 3 above)
Comment (jplevyak) [reply content/en/docs/concepts/wasm/index.md]: Tried to clean that up.  PTAL

File: content/en/docs/concepts/wasm/index.md (hunk 2 above)
Comment (jplevyak) [reply content/en/docs/concepts/wasm/index.md]: Done.

File: content/en/docs/concepts/wasm/index.md (hunk 1 above)
Comment (jplevyak) [reply content/en/docs/concepts/wasm/index.md]: Done.

File: content/en/docs/concepts/wasm/index.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
+---
+title: WebAssembly (Wasm) Plugins
+description: Describes Istio's WebAssembly Plugin system.
+weight: 25
+keywords: [wasm,webassembly,emscripten,extension,plugin,filter]
+---
+
+WebAssembly is a sandboxing technology which can be used to extend the Istio proxy (Envoy).  Going forward Istio extensions will be written as proxy-Wasm plugins using the WebAssembly (Wasm) sandbox API. Mixer will be eliminated from the Istio architecture and all Mixer adapters must be eventually rewritten as proxy plugins.
```
Comment (ericvn) [inline content/en/docs/concepts/wasm/index.md]: last sentence ends in proxy plugin. Should it be proxy-Wasm plugin?

File: content/en/docs/concepts/wasm/index.md [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+- **Call out APIs** for gRPC and HTTP calls
+
+- **Stats and Logging APIs** for metrics and monitoring
+
+{{< image width="80%" link="./extending.svg" caption="Extending Istio/Envoy" >}}
+
+## Example
+
+An example C++ Wasm proxy-Wasm plugin for a filter can be found
```
Comment (ericvn) [inline content/en/docs/concepts/wasm/index.md]: Is the extra `Wasm` supposed to be here?

File: content/en/docs/concepts/wasm/index.md [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+keywords: [wasm,webassembly,emscripten,extension,plugin,filter]
+---
+
+WebAssembly is a sandboxing technology which can be used to extend the Istio proxy (Envoy).  Going forward Istio extensions will be written as proxy-Wasm plugins using the WebAssembly (Wasm) sandbox API. Mixer will be eliminated from the Istio architecture and all Mixer adapters must be eventually rewritten as proxy plugins.  
+
+WebAssembly sandbox goals:
+
+- **Efficiency** - An extension adds low latency, CPU, and memory overhead. 
+- **Function** - An extension can enforce policy, collect telemetry, and perform payload mutations.
+- **Isolation** - A programming error or crash in an extension does affect other plugin.
```
Comment (douglas-reid) [inline content/en/docs/concepts/wasm/index.md]: s/plugin/extensions/

File: content/en/docs/concepts/wasm/index.md (hunk 6 above)
Comment (mandarjog) [inline content/en/docs/concepts/wasm/index.md]:
```suggestion
WebAssembly is a sandboxing technology which can be used to extend the Istio proxy (Envoy).  The Proxy-Wasm sandbox API replaces Mixer as the primary extension mechanism in Istio. Istio 1.6 will provide a uniform configuration API for Proxy-Wasm plugins.
```

Comment (mandarjog) [reply content/en/docs/concepts/wasm/index.md]: Can we add a small paragraph before this? "What is WebAssembly?" with a link to webassembly.org or some other authoritative source.

**Comment by duderino** [comment]
@jplevyak 1.5 will be published today.  I think this one is not a blocker so we'll skip it unless you can get it quickly

**Comment by PiotrSikora** [comment]
John is already OOO, so either me or @mandarjog can re-open this as a new PR, and we can iterate there.

File: content/en/docs/concepts/wasm/index.md (hunk 3 above)
Comment (louiscryan) [inline content/en/docs/concepts/wasm/index.md]: Lets link to other community resources

Comment (PiotrSikora) [reply content/en/docs/concepts/wasm/index.md]: Done, PTAL.

File: content/en/docs/concepts/wasm/index.md [hunk 9]
```diff
@@ -0,0 +1,56 @@
+---
+title: WebAssembly (Wasm) Plugins
```
Comment (louiscryan) [inline content/en/docs/concepts/wasm/index.md]: WASM isn't a concept - its a technology.

The 'concept' here is 'Extensibility' can we retitle ?

**Comment by PiotrSikora** [comment]
Damn it, someone broke website and linter complains about unrelated change, even after merging `master`. Can we force-merge?

**Pull request merged**